h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java

```diff
@@ -440,7 +446,24 @@ protected XGBoostModel deepClone(Key<XGBoostModel> result) {
   }
   
   private static int getMaxNThread() {
-    return Integer.getInteger(SYSTEM_PROP_PREFIX + "xgboost.nthread", H2O.ARGS.nthreads);
+    if (System.getProperty(PROP_NTHREAD) != null) {
+      return Integer.getInteger(System.getProperty(PROP_NTHREAD));
+    } else {
+      int maxNodesPerHost = 1;
+      Set<String> checkedNodes = new HashSet<>();
+      for (H2ONode node : H2O.CLOUD.members()) {
+        String nodeHost = node.getIpHostString();
+        if (!checkedNodes.contains(nodeHost)) {
+          checkedNodes.add(nodeHost);
+          long cnt = Stream.of(H2O.CLOUD.members()).filter(h -> h.getIpHostString().equals(nodeHost)).count();
+          if (cnt > maxNodesPerHost) {
+            maxNodesPerHost = (int) cnt;
+          }
+        }
+      }
+      int maxThreads = H2O.ARGS.nthreads;
+      return Math.max(1, maxThreads / maxNodesPerHost);
```

Comment:
it's probably a corner case or a bad practice to use multiple nodes per host (outside testing), but why dividing by `maxNodesPerHost` and not by `nodesOnCurrentHost`? is it just to ensure that all the nodes have the same config, i.e. same `nthreads`?

Reply:
the problem is that with hadoop this can happen quite easily
the config is global so thats why maxNodesPerHost, but the more I think about this I think it can be improved, thanks for the nudge

Reply:
The issue here is that it won't most likely make anything faster because the weakest link will always slow down the others in the all-reduce setting.

I think keeping the cluster uniform is maybe better, memory allocation will be more predictable. You can make it configurable and see what is behaving better.

Reply:
You don't see the other cluster and what they are doing, however, the single worst case gives you a glimpse of what might be going on the other nodes. For deployments where the resource sharing could happen - I would prefer stability over (a marginal) gain of speed. If the gain is significant - that is another question :)